chain-core/src/property.rs

```diff
@@ -0,0 +1,228 @@
+    /// absolute position of the block in the chain. This can be used
+    /// for random access (if the storage algorithm allows it) or for
+    /// identifying the position of a block in a given epoch or era.
+    type Date;
+
+    /// return the Block's identifier.
+    fn id(&self) -> Self::Id;
+
+    /// get the parent block identifier (the previous block in the
+    /// blockchain).
+    fn parent_id(&self) -> &Self::Id;
+
+    /// get the block date of the block
+    fn date(&self) -> &Self::Date;
+}
+
+/// define a transaction within the blockchain. This transaction can be used
+/// for the UTxO model. However it can also be used for any other elements that
+/// the blockchain has (a transaction type to add Stacking Pools and so on...).
+///
+pub trait Transaction: Serializable {
+    /// the input type of the transaction (if none use `()`).
+    type Input;
+    /// the output type of the transaction (if none use `()`).
+    type Output;
+    /// a unique identifier of the transaction. For 2 different transactions
+    /// we must have 2 different `Id` values.
+    type Id;
+
+    fn inputs<'a>(&'a self) -> std::slice::Iter<'a, Self::Input>;
+    fn outputs<'a>(&'a self) -> std::slice::Iter<'a, Self::Output>;
+
+    /// return the Transaction's identifier.
+    fn id(&self) -> Self::Id;
+}
+
+/// accessor to transactions within a block
+///
+/// This trait is generic enough to show there is multiple types
+/// of transaction possibles:
+///
+/// * UTxO
+/// * certificate registrations
+/// * ...
+pub trait HasTransaction<T: Transaction> {
+    /// returns an iterator over the Transactions
+    fn transactions<'a>(&'a self) -> std::slice::Iter<'a, T>;
+}
+
+/// Define the Ledger side of the blockchain. This is not really on the blockchain
+/// but should be able to maintain a valid state of the overall blockchain at a given
+/// `Block`.
+pub trait Ledger<T: Transaction> {
+    /// a Ledger Update. An atomic representation of a set of changes
+    /// into the ledger's state.
+    ///
+    /// This can be seen like a git Diff where we can see what is going
+    /// to be removed from the Ledger state and what is going to be added.
+    type Update;
+
+    /// Ledger's errors
+    type Error: std::error::Error;
+
+    /// check the input exists in the given ledger state
+    ///
+    /// i.e. in the UTxO model the Input will be something like the Transaction's Id
+    /// and the index of the output within the output array.
+    /// If the Output is not present it is possible that it does not exist or has
+    /// already been spent in another transaction.
+    fn input<'a>(&'a self, input: &T::Input) -> Result<&'a T::Output, Self::Error>;
+
+    /// create a new Update from the given transaction.
+    fn diff_transaction(&self, transaction: &T) -> Result<Self::Update, Self::Error>;
+
+    /// create a combined Update from the given transactions
+    fn diff<'a, I>(&self, transactions: I) -> Result<Self::Update, Self::Error>
```

Comment:
This could probably have a default implementation in terms of `diff_transaction` method.

Reply:
indeed, the only constraint I see is that we will need to add something on the `Self::Update` type.

```rust
    type Update: std::ops::Add<Self::Update>;
```

Reply:
Ah yes, there would need to be a generic reduction op indeed.